File: problems/009-sparse-file-quota/solutions/python/main.py
import sys

t = sys.stdin.buffer.read().split()
it = iter(t)
F, N, B = map(int, (next(it), next(it), next(it)))
size = [0] * (F + 1)
cur = [0] * (F + 1)
used = peak = 0
out = []
for _ in range(N):
    op = next(it)
    x = int(next(it))
    v = int(next(it))
    err = False
    if op == b"SEEK":
        cur[x] = v
    else:
        ns = (size[x] if v == 0 else max(size[x], cur[x] + v)) if op == b"WRITE" else v
        if ns > size[x] and ns - size[x] > B - used:
            err = True
        else:
            used += ns - size[x]
            size[x] = ns
            if op == b"WRITE" and v:
                cur[x] += v
    peak = max(peak, used)
    out.append(("ERR QUOTA" if err else "OK") + f" {size[x]} {cur[x]} {used}")
out.append(f"SUMMARY {used} {peak}")
print("\n".join(out))

File: problems/009-sparse-file-quota/solutions/rust/main.rs
use std::io::{self, Read};
fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();
    let mut t = s.split_whitespace();
    let f: usize = t.next().unwrap().parse().unwrap();
    let n: usize = t.next().unwrap().parse().unwrap();
    let cap: u64 = t.next().unwrap().parse().unwrap();
    let mut size = vec![0u64; f + 1];
    let mut cur = vec![0u64; f + 1];
    let (mut used, mut peak) = (0u64, 0u64);
    let mut out = String::new();
    for _ in 0..n {
        let op = t.next().unwrap();
        let x: usize = t.next().unwrap().parse().unwrap();
        let v: u64 = t.next().unwrap().parse().unwrap();
        let mut err = false;
        if op == "SEEK" {
            cur[x] = v
        } else {
            let ns = if op == "WRITE" {
                if v == 0 {
                    size[x]
                } else {
                    size[x].max(cur[x] + v)
                }
            } else {
                v
            };
            if ns > size[x] && ns - size[x] > cap - used {
                err = true
            } else {
                if ns >= size[x] {
                    used += ns - size[x]
                } else {
                    used -= size[x] - ns
                }
                size[x] = ns;
                if op == "WRITE" && v > 0 {
                    cur[x] += v
                }
            }
        }
        peak = peak.max(used);
        out += &format!(
            "{} {} {} {}\n",
            if err { "ERR QUOTA" } else { "OK" },
            size[x],
            cur[x],
            used
        );
    }
    out += &format!("SUMMARY {} {}\n", used, peak);
    print!("{}", out);
}
